Document Pomerium Policy Language (#2784)
pomerium/pomerium: docs/docs/topics/ppl.md

---
title: Pomerium Policy Language
description: >-
  This article covers Pomerium Policy Language, used to define secure access policies for routes.
---

# Pomerium Policy Language

> Pomerium Policy Language (**PPL**) is a yaml-based notation for creating easy and flexible authorization policies. This document covers the usage of PPL and provides several example policies.

Reviewer: ```suggestion
Pomerium Policy Language (**PPL**) is a [yaml](https://en.wikipedia.org/wiki/YAML)-based notation for creating easy and flexible authorization policies. This document covers the usage of PPL and provides several example policies.
```

Author: ```suggestion
Pomerium Policy Language (**PPL**) is a [yaml]-based notation for creating easy and flexible authorization policies. This document covers the usage of PPL and provides several example policies.
```

## At a Glance

> Each PPL policy has at the top level a set of `allow` or `deny` actions, with a list of logitical operators, criteria, matchers, and values underdeath. For example:

Reviewer 2: ```suggestion
Each PPL policy has at the top level a set of `allow` or `deny` actions, with a list of logical operators, criteria, matchers, and values underneath. For example:
```

```yaml
allow:
  and:
  - domain: 
      is: example.com
  - groups: 
      has: admin
deny:
  or:
  - user: 
      is: user1@example.com
  - user: 
      is: user2@example.com
```

> This policy will allow a user with an email address at `example.com` who **is also** a member of the `admin` group. It will deny `user1` or `user2`, regardless of their domain and group membership.

Reviewer: ```suggestion
This policy will allow a user with an email address at `example.com` who **is also** a member of the `admin` group. It will deny `user1` **or** `user2`, regardless of their domain and group membership.
```

```yaml
allow:
  and:
  - claim/family_name: Smith
```

PPL supports many different criteria:

| Criterion Name               | Data Format                   | Description                                                                                                                                                                                              |
| ---------------------------- | ----------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `accept`                     | Anything. Typically `true`.   | Always returns true, thus always allowing access. Equivalent to the [`allow_public_unauthenticated_access`] option.                                                                                      |
| `authenticated_user`         | Anything. Typically `true`.   | Always returns true for logged-in users. Equivalent to the [`allow_any_authenticated_user`] option.                                                                                                      |

> | `claim`                      | Anything. Typically a string. | Returns true if a token claim matches the supplied value **exactly**. The claim to check is determined via the sub-path. <br/> For exampe, `claim/family_name: Smith` matches if the user's `family_name` claim is `Smith`. |

Reviewer: ```suggestion
| `claim`                      | Anything. Typically a string. | Returns true if a token claim matches the supplied value **exactly**. The claim to check is determined via the sub-path. <br/> For example, `claim/family_name: Smith` matches if the user's `family_name` claim is `Smith`. |
```

| Criterion Name | Data Format         | Description                                                                            |
| -------------- | ------------------- | -------------------------------------------------------------------------------------- |
| `date`         | Date Matcher        | Returns true if the time of the request matches the constraints.                       |
| `day_of_week`  | Day of Week Matcher | Returns true if the day of the request matches the constraints.                        |
| `time_of_day`  | Time of Day Matcher | Returns true if the time of the request (for the current day) matches the constraints. |

## Matchers

### String Matcher

A string matcher is an object with operators as keys. It supports the following operators: `contains`, `ends_with`, `is` and `starts_with`. For example:

```yaml
allow:
  and:
  - email:
      starts_with: 'admin@'
```

### List Matcher

A list matcher is an object with operators as keys. It supports the following operators: `has`. For example:

```yaml
allow:
  and:
  - groups:
      has: 'admin'
```

## Date Matcher

> The date matcher is an object with operators as keys. It supports the following operators: `after` and `before`. The values are ISO-8601 date strings. `after` means that the time of the request must be after the supplied date and `before` means that the time of the request must be before the supplied date. For example:

Reviewer 2: ```suggestion
The date matcher is an object with operators as keys. It supports the following operators: `after` and `before`. The values are [ISO-8601](https://en.wikipedia.org/wiki/ISO_8601) date strings. `after` means that the time of the request must be after the supplied date and `before` means that the time of the request must be before the supplied date. For example:
```

    ```yaml
    allow:
      and:
        - day_of_week: tue-fri
    ```

## Time of Day Matcher

The time of day matcher is an object with operators as keys. It supports the following operators: `timezone`, `after`, and `before`.

`timezone` is required and specifies the timezone to use when interpreting the supplied times. It is recommended to use city names (like `America/Phoenix`) instead of standard timezone abbreviations because standard timezones change throughout the year (i.e. EST becomes EDT and back again).
 
`after` means the time of the request must be after the supplied time and `before` means that the time of the request must be before the supplied time. For example:

```yaml
allow:
  and:
    - time_of_day:
        timezone: UTC
        after: 2:20:00
        before: 4:30PM
```

[`allow_public_unauthenticated_access`]: /reference/readme.md#public-access
[`allow_any_authenticated_user`]: /reference/readme.md#allow-any-authenticated-user
[CORS pre-flight requests]: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#preflighted_requests
[client certificate authority]: /reference/readme.md#client-certificate-authority

> [Pomerium Enterprise]: /enterprise/about.md

Author: ```suggestion
[Pomerium Enterprise]: /enterprise/about.md
[yaml]: https://en.wikipedia.org/wiki/YAML

```
